Issue/personal keys

bharath-orchestral opened the pull request:
Code to add feature where Users(non-admin) can manage their own keys. 
Added a key in config file(personal_keys under rbac). When the key is enabled, Users(non-admin) can create/update/delete/view their own keys and cannot create/update/delete/view other user's keys.

If the 'personal_keys' key is disabled everything behaves as before. Nothing is changed in regards to admins.

bharath-orchestral commented:
Can someone help me on 'ci/circleci:packages' and 'Orquesta CI/Integration Tests' errors? And also where to add changes in CHANGELOG.rst?

amanda11 reviewed (changes requested):
Some minor comments but in addition:
1. It would be good if the PR could mention the use-case for this
2. Need to add changelog entry - needs to go into the in development section, in the added section. Just a new bullet that matches the others in there for formatting
3. Not sure why the init.py from dummy_pack_23 has been removed
4. I think it needs some more tests to verify behaviour depending on value of personal_keys and whether the rbac backend allows user or not.
5. I think we need to consider the cases where RBAC is not enabled. That may just be a reverse of some logic you have added, but I've not considered it deeply - just think that it is missing
6. In general, then should this be a global setting. Why is it a global setting as to whether personal key is amended. Perhaps that would be explained by more information in the PR, but need to understand use-case and why it has been added as a global setting rather than say another RBAC permission that is given to individual users.
 
on the failres. I've re-run the failing integration job and the CircleCI jobs, let's see if they pass. As it didn'nt seem to be related to the code changes.

nzlosh commented on st2api/st2api/controllers/v1/auth.py:
```diff
@@ -126,18 +136,42 @@ def get_all(self, requester_user, show_secrets=None, limit=None, offset=0):
+        rbac_utils = get_rbac_backend().get_utils_class()
+        user_is_admin = rbac_utils.user_is_admin(user_db=requester_user)
+        is_same_user = requester_user.name == api_key_api.user
+        if (user_is_admin) or (is_same_user or api_key_api.user == ""):
+            return True
+        else:
+            return False
```
Why apply a conditional test here when the evaluation result can be returned directly?

```python
return (user_is_admin) or (is_same_user or api_key_api.user == "")
```

bharath-orchestral commented:
@amanda11 
I have added some more context in the PR. 
Added changes in CHANGELOG file.
Made the flag false by default
Added __init__.py file. Don't know why it got removed in the first place
@nzlosh 
Made the suggested code change

bharath-orchestral commented on st2api/st2api/controllers/v1/auth.py:
```diff
@@ -233,11 +277,21 @@ def delete(self, api_key_id_or_key, requester_user):
-            resource_db=api_key_db,
-            permission_type=permission_type,
-        )
+        if cfg.CONF.rbac.personal_keys:
```
Can you help me as to where to add the code? I cannot understand where the rbac_utils methods are implemented.I have checked base.py in st2common/st2common/rbac/backends/ . Do I need to add the method there, check if conf.rbac.enabled is true and proceed?

amanda11 replied:
I was more suggesting that you reverse the order in this method. e.g. First call rbac_utils method to see if they have permission to access database. Then if the rbac_utils calls says No, you add in your check for personal keys.
That way when the noop backend is used (which will always say yes to permission), you don't need to check personal keys - and can keep personal keys default to false.

There are multiple rbac backends, so if you wanted to add a new method you would need to add into all implementations. The base.py is just the base rbac class which always says not implemented. The noop.py is a always yes utils. Then in the st2-rbac-backend GIT repo you will find the implementation that is used when you really do use RBAC - which uses the rbac roles and auth files.

amanda11 commented:
In general, I'd like some consideration from the @StackStorm/tsc  as to whether it makes sense to keep this as is - eg. global option that allows if set, so that all users can manage their own keys. OR if there is a use-case where you want to make it an RBAC permission, such that some users can create their own keys but others can't. For example, if we are adding the ability to give permission to manage your own api keys, would it be better to be added as a RBAC permission, e.g. USER_API_KEY_VIEW etc - which allows user to manage their own keys. That way, you can allow some non-admins to manage their keys, but others not. This would be more flexible, and more in keeping with having the permissions scheme we already have. For example, if you want all non-admins to have that, then you'd give that permission to a role that all users had.

I'm just not sure about starting trend of having some permissions controlled by role, and others by system flags. Think it might complicate things long term, and the change to introduce a new permission in my view would be more in sync with current system.

But I'd welcome other TSC members views.

nzlosh commented:
I agree with the points @amanda11 has made to keep the feature in the RBAC security model.  It's unfortunate the user ownership feature can't be used directly https://docs.stackstorm.com/rbac.html#restricting-users-to-only-view-resources-they-own-or-created to allow users to manage their own keys but perhaps it could be used to inspire an implementation.

armab commented:
@bharath-orchestral Fine-grained control for users to manage their own API keys is an excellent addition! It would be nice to have it in the st2 core :+1: 

I totally agree about the interface and implementation with @amanda11 and @nzlosh that it's best to have this functionality as a new RBAC permission, instead of the `st2.conf` option.

bishopbm1 commented:
I agree with @amanda11 also. I think this fits the bill more in RBAC then in the st2.conf.

We are seeing more and more people turn on RBAC controls to help manage users coming into the platform and using it. I think this falls right inline with that. I like @nzlosh idea about the ownership feature and would be a great implementation as well.

amanda11 commented:
Thanks for the updates... @bharath-orchestral I think its worth taking a look at the ownership feature that @nzlosh mentioned, and also take a look at the st2-rbac-backend GIT repo, in there you'll see in there a ./st2rbac_backend/utils.py which is where the current checking on the role permissions is done. So if its implemented as a new role permission then it would be around there that would be added.
For the ownership feature then I'm not sure how that's implemented, to advise how easy that would be. But feel free to ask on here, or on the StackStorm thread to help, and I'm sure one of us from the @StackStorm/tsc can help answer any queries.

It would be great to get this feature in, but consensus seems to be agreed to try and move this from the st2.conf into a proper RBAC feature. Hope, this has given some guidance of how to move that into the main RBAC, but we'll help where we can to assist you moving this.
